pageserver/src/http/routes.rs

```diff
@@ -0,0 +1,164 @@
+use std::sync::Arc;
+
+use anyhow::Result;
+use hyper::header;
+use hyper::StatusCode;
+use hyper::{Body, Request, Response, Uri};
+use routerify::Middleware;
+use routerify::{ext::RequestExt, RouterBuilder};
+use zenith_utils::auth::JwtAuth;
+use zenith_utils::http::endpoint::attach_openapi_ui;
+use zenith_utils::http::endpoint::auth_middleware;
+use zenith_utils::http::endpoint::check_permission;
+use zenith_utils::http::endpoint::AuthProvider;
+use zenith_utils::http::error::ApiError;
+use zenith_utils::http::{
+    endpoint,
+    error::HttpErrorBody,
+    json::{json_request, json_response},
+};
+
+use super::models::BranchCreateRequest;
+use super::models::TenantCreateRequest;
+use crate::page_cache;
+use crate::{
+    branches::{self},
+    PageServerConf, ZTenantId,
+};
+
+#[derive(Debug)]
+struct State {
+    conf: &'static PageServerConf,
+    auth: Arc<Option<JwtAuth>>,
+    whitelist_routes: Vec<Uri>,
+}
+
+impl State {
+    fn new(conf: &'static PageServerConf, auth: Arc<Option<JwtAuth>>) -> Self {
+        let whitelist_routes = ["/v1/status", "/v1/doc", "/swagger.yml"]
+            .iter()
+            .map(|v| v.parse().unwrap())
+            .collect::<Vec<_>>();
+        Self {
+            conf,
+            auth,
+            whitelist_routes,
+        }
+    }
+}
+
+impl AuthProvider for State {
+    fn provide_auth(&self, req: &Request<Body>) -> Arc<Option<JwtAuth>> {
+        if self.whitelist_routes.contains(req.uri()) {
+            Arc::new(None)
+        } else {
+            self.auth.clone()
+        }
+    }
```

Comment:
This is another place where the unfortunate choice of types forces us to do a `.clone()`. Here's one way to fix this:

```rust
fn provide_auth(&self, req: &Request<Body>) -> Option<&JwtAuth> {
    if self.whitelist_routes.contains(req.uri()) {
        None
    } else {
        Option::as_ref(&self.auth)
    }
}
```